Adding FLUSHING state to tasks and stages

rohangarg opened the pull request:
Adds a flushing state to tasks and stages.
This is useful in case of broadcast joins where if the build source stage is done processing and is just running to serving data from its output buffer to join tasks, its sub-stages can be shutdown by cancellation (since the new output produced by the sub stages will not be useful). The cancellation is helpful since the sub stages under build source can be complex and take significant memory (the output buffer size is configurable) and CPU (due to the processing) to get to a blocked state.

For tasks, flushing state signifies that there will be no new drivers, the existing drivers have finished and the output buffer of the task is at-least in a 'no-more-pages' state.
For stages, flushing state signifies that at-least one of its task is flushing and the non-flushing tasks are finished.
For both, once the state has started flushing it will never go back to 'RUNNING' state.

findepi commented:
> The cancellation is helpful since the sub stages under build source can be complex and take significant memory (the output buffer size is configurable) and CPU (due to the processing) to get to a blocked state.

If the build side completed building the lookup, the source is finished, otherwise the build side would not know the lookup is "full".
If there are computations lingering in the lower stages, this seems not specific to join build side.

To put differently, I do not understand the situation this is helping with.

rohangarg commented:
>If the build side completed building the lookup, the source is finished, otherwise the build side would not know the lookup is "full".
If there are computations lingering in the lower stages, this seems not specific to join build side.
To put differently, I do not understand the situation this is helping with.

consider a broadcast join scenario where the build side is very small (1KB) and is computed and stored in build source's output buffer. Being a broadcast buffer, it will keep running until it is told that no new tasks will be added in the join stage - since it would have to serve the new join tasks if created. There are scenarios where it will never be told that no new tasks are coming (that logic will be improved as well in future changes). Now since the build source is running, its sub stages will also keep on running until they get blocked (because the build source is not consuming their output). The time and memory taken to reach to blocking state can be significant.

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/StageState.java:
```diff
@@ -43,6 +43,10 @@
      */
     RUNNING(false, false),
+    /**
+     * Stage has finished executing and output being consumed.
```
I would also add a sentence from a commit message:
```
(at-least one of its tasks is flushing and the non-flushing tasks are finished).
```

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/StageState.java:
```diff
@@ -112,4 +116,9 @@ public boolean canScheduleMoreTasks()
         return true;
     }
+
+    public boolean isRunning()
```
Rename it to:
```
isRunningOrFlushing
```

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/SqlStageExecution.java:
```diff
@@ -524,6 +536,12 @@ else if (taskState == TaskState.FINISHED) {
         }
     }
 
+    private boolean isFlushing()
```
make it synchronized

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/TaskState.java:
```diff
@@ -64,4 +68,9 @@ public boolean isDone()
         return doneState;
     }
+
+    public boolean isRunning()
```
rename to `isRunningOrFlushing`

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/TaskState.java:
```diff
@@ -30,6 +30,10 @@
      */
     RUNNING(false),
+    /**
+     * Task has finished executing and output is left to be consumed
```
add a sentence from PR description:
```
Flushing state signifies that there will be no new drivers,
the existing drivers have finished and the output buffer of
the task is at-least in a 'no-more-pages' state.
```

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/MemoryRevokingScheduler.java:
```diff
@@ -245,7 +245,7 @@ private void requestRevoking(MemoryPool memoryPool, Collection<SqlTask> sqlTasks
         AtomicLong remainingBytesToRevokeAtomic = new AtomicLong(remainingBytesToRevoke);
         sqlTasks.stream()
-                .filter(task -> task.getTaskStatus().getState() == TaskState.RUNNING)
+                .filter(task -> task.getTaskStatus().getState().isRunning())
```
We shouldn't request revoking from `FLUSHING` task as no drivers are running which could revoke memory

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/MemoryRevokingScheduler.java:
```diff
@@ -218,7 +218,7 @@ private boolean memoryRevokingNeeded(MemoryPool memoryPool)
     {
         return sqlTasks.stream()
-                .filter(task -> task.getTaskStatus().getState() == TaskState.RUNNING)
+                .filter(task -> task.getTaskStatus().getState().isRunning())
```
We can keep `== TaskState.RUNNING` as `FLUSHING` tasks have no memory to revoke

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/StageStateMachine.java:
```diff
@@ -253,7 +259,7 @@ public BasicStageStats getBasicStageStats(Supplier<Iterable<TaskInfo>> taskInfos
         // never be visible.
         StageState state = stageState.get();
-        boolean isScheduled = (state == RUNNING) || state.isDone();
+        boolean isScheduled = state.isRunning() || state.isDone();
```
Maybe we should have `StageState#isScheduled` method instead.

rohangarg replied:
There is already a `scheduled` state for a stage - so having `isScheduled` could lead to ambiguity. Although, by that logic - the variable can be named differently (like `begunExecution`)

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTask.java:
```diff
@@ -114,10 +114,10 @@ public void testEmptyQuery()
                         .withNoMoreBufferIds(),
                 OptionalInt.empty());
-        assertEquals(taskInfo.getTaskStatus().getState(), TaskState.RUNNING);
+        assertTrue(taskInfo.getTaskStatus().getState().isRunning());
```
please test actual states and don't use `isRunning()`. We need to have test where we validate `FLUSHING` state

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -177,7 +178,7 @@ public void testSimple(PipelineExecutionStrategy executionStrategy)
             //
             // test body
-            assertEquals(taskStateMachine.getState(), TaskState.RUNNING);
+            assertTrue(taskStateMachine.getState().isRunning());
```
please test actual states and don't use `isRunning()`. We need to have test where we validate `FLUSHING` state

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -279,6 +280,9 @@ public void testSimple(PipelineExecutionStrategy executionStrategy)
 
             outputBufferConsumer.abort(); // complete the task by calling abort on it
             TaskState taskState = taskStateMachine.getStateChange(TaskState.RUNNING).get(10, SECONDS);
+            if (taskState == TaskState.FLUSHING) {
```
We should incorporate check for `FLUSHING` state into test scenario

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -419,7 +420,7 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
         Set<SqlStageExecution> childStages = childStagesBuilder.build();
         stage.addStateChangeListener(newState -> {
-            if (newState.isDone()) {
+            if (newState.isDone() || newState == FLUSHING) {
```
Please add a test for this. The test could be:
1. staring `SELECT * FROM large_table LIMIT 1`
2. NOT fetching results and waiting for SOURCE stage to be cancelled

rohangarg commented:
@sopel39 : Thank you for the review! I have addressed all the comments w.r.t non-test code. Also, I have made the `Task` level tests more specific in terms of state asserting and incorporated exact testing for `FLUSHING` state.
I'll also add more tests for `Stage` level flushing asap too.

dain commented on presto-main/src/main/java/io/prestosql/execution/SqlStageExecution.java:
```diff
@@ -524,6 +536,12 @@ else if (taskState == TaskState.FINISHED) {
 
+    private synchronized boolean isFlushing()
+    {
+        return flushingTasks.size() > 0
```
Add a comment explaining that to be flushing, there must be at least one flushing task, and all others must be flushing or finished.

also replace `flushingTasks.size() > 0` with `!flushingTasks.isEmpty()`

dain commented on presto-main/src/main/java/io/prestosql/execution/TaskState.java:
```diff
@@ -64,4 +70,9 @@ public boolean isDone()
         return doneState;
     }
+
+    public boolean isRunningOrFlushing()
```
I'm not a fan of this method.  In all uses, it seem to be used in an `||` with some other clause on the state.  Instead, I would just inline this method.

rohangarg commented:
@dain : Thanks for the review!

I have incorporated the current review comments and added a test for the case 
>Please add a test for this. The test could be:
SELECT * FROM large_table LIMIT 1
NOT fetching results and waiting for SOURCE stage to be cancelled

Will non query based UTs for `SqlStageExecution` be useful too? I tried to write it following the test present in `TestSqlStageExecution` in presto-main - but faced some difficulty with controlling state transitions for a stage. I can retry it if the tests seem to be needed.

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -278,7 +282,13 @@ public void testSimple(PipelineExecutionStrategy executionStrategy)
             }
 
             outputBufferConsumer.abort(); // complete the task by calling abort on it
-            TaskState taskState = taskStateMachine.getStateChange(TaskState.RUNNING).get(10, SECONDS);
```
I think we can simplify it by:
```
TaskState taskState = taskStateMachine.getStateChange(taskStateMachine.getState()).get(10, SECONDS);
```
not that if `getState` returns `FINISHED`, future will complete immediately.

rohangarg replied:
I wanted to do that itself - but the `getStateChange` method throws an exception if the `getState` returns `FINISHED`.
Also, if the state is running, the function might return if the task transitions to `FLUSHING`. And we'd have to run `getStateChange` again.

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -428,7 +438,7 @@ public void testComplex(PipelineExecutionStrategy executionStrategy)
             //
             // test body
-            assertEquals(taskStateMachine.getState(), TaskState.RUNNING);
+            assertTrue(taskStateMachine.getState() == RUNNING || taskStateMachine.getState() == FLUSHING);
```
why we cannot assert:
```
assertEquals(taskStateMachine.getState(), RUNNING);
```
as in `testSimple`?

rohangarg replied:
Yes you are right, I missed that change.

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -474,6 +484,7 @@ public void testComplex(PipelineExecutionStrategy executionStrategy)
                     // assert that task result is produced
                     outputBufferConsumer.consume(100 * 5 * 3, ASSERT_WAIT_TIMEOUT);
                     outputBufferConsumer.assertBufferComplete(ASSERT_WAIT_TIMEOUT);
+                    assertEquals(taskStateMachine.getStateChange(RUNNING).get(10, SECONDS), FLUSHING);
```
this check doesn't make sense if we assert:
```
assertTrue(taskStateMachine.getState() == RUNNING || taskStateMachine.getState() == FLUSHING);
```
above

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -580,7 +591,13 @@ public void testComplex(PipelineExecutionStrategy executionStrategy)
             }
 
             outputBufferConsumer.abort(); // complete the task by calling abort on it
-            TaskState taskState = taskStateMachine.getStateChange(TaskState.RUNNING).get(10, SECONDS);
```
I think we can simplify it by:
```
TaskState taskState = taskStateMachine.getStateChange(taskStateMachine.getState()).get(10, SECONDS);
```
not that if `getState` returns `FINISHED`, future will complete immediately.

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskManager.java:
```diff
@@ -111,20 +113,19 @@ public void testSimpleQuery()
 
-            taskInfo = sqlTaskManager.getTaskInfo(taskId);
-            assertEquals(taskInfo.getTaskStatus().getState(), TaskState.RUNNING);
+            taskInfo = sqlTaskManager.getTaskInfo(taskId, taskInfo.getTaskStatus().getState()).get(1, TimeUnit.SECONDS);
```
this will wait for 1s if `taskInfo.getTaskStatus().getState()` is already FLUSHING. You could use:
```
sqlTaskManager.getTaskInfo(taskId, RUNNING).get(1, TimeUnit.SECONDS);
```
instead

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskManager.java:
```diff
@@ -191,10 +192,9 @@ public void testAbortResults()
 
-            taskInfo = sqlTaskManager.getTaskInfo(taskId);
-            assertEquals(taskInfo.getTaskStatus().getState(), TaskState.RUNNING);
+            taskInfo = sqlTaskManager.getTaskInfo(taskId, taskInfo.getTaskStatus().getState()).get(1, TimeUnit.SECONDS);
```
ditto

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -421,7 +421,7 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
         }
         Set<SqlStageExecution> childStages = childStagesBuilder.build();
         stage.addStateChangeListener(newState -> {
-            if (newState.isDone()) {
```
change the order of conditions:
```
if (newState == FLUSHING || newState.isDone())
```

sopel39 commented on presto-tests/src/test/java/io/prestosql/execution/TestSqlStageExecution.java:
```diff
@@ -0,0 +1,65 @@
+    public void testFlushingState()
+            throws Exception
+    {
+        queryRunner = createQueryRunner(ImmutableMap.of("node-scheduler.include-coordinator", "false"));
```
Do we need distributed runner for this test?

rohangarg replied:
No, we don't need a distributed runner for this test. Used it because `TestQueryRunnerUtil` provided a clean interface for `waitForQueryState` (without fetching results) - is any existing one with `LocalQueryRunner` too?

sopel39 commented on presto-tests/src/test/java/io/prestosql/execution/TestSqlStageExecution.java:
```diff
@@ -0,0 +1,65 @@
+
+        QueryInfo queryInfo = queryRunner.getCoordinator().getFullQueryInfo(queryId);
+        while (queryInfo.getOutputStage().get().getState() != FLUSHING) {
+            queryInfo = queryRunner.getCoordinator().getFullQueryInfo(queryId); // wait for the query to finish producing results, but don't poll them
```
put some sleep here to prevent active loop, e.g:
```
            MILLISECONDS.sleep(100);
```

sopel39 commented on presto-tests/src/test/java/io/prestosql/execution/TestSqlStageExecution.java:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        while (queryInfo.getOutputStage().get().getSubStages().get(0).getState() != CANCELED) {
+            queryInfo = queryRunner.getCoordinator().getFullQueryInfo(queryId); // wait for the sub stages to go to cancelled state
```
dittto

sopel39 commented on presto-tests/src/test/java/io/prestosql/execution/TestSqlStageExecution.java:
```diff
@@ -0,0 +1,65 @@
+        assertEquals(queryInfo.getState(), RUNNING);
+        assertEquals(queryInfo.getOutputStage().get().getState(), FLUSHING);
+        assertEquals(queryInfo.getOutputStage().get().getSubStages().size(), 1);
+        assertEquals(queryInfo.getOutputStage().get().getSubStages().get(0).getState(), CANCELED);
```
could you also pull results here and wait for the query to finish gracefully?

sopel39 commented:
tests are failing (I've seen some checkstyle issues)

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/StageState.java:
```diff
@@ -43,6 +43,11 @@
      * Stage is running.
      */
     RUNNING(false, false),
+    /**
```
please update `canScheduleMoreTasks` method

rohangarg commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -177,7 +180,7 @@ public void testSimple(PipelineExecutionStrategy executionStrategy)
 
             //
             // test body
-            assertEquals(taskStateMachine.getState(), TaskState.RUNNING);
```
added that as a separate commit

sopel39 commented on presto-main/src/main/java/io/prestosql/execution/SqlTaskExecution.java:
```diff
@@ -641,6 +641,7 @@ private synchronized void checkTaskCompletion()
 
         // are there still pages in the output buffer
         if (!outputBuffer.isFinished()) {
```
@dain is this a valid transition for grouped execution or there should be additional conditions? I'm thinking that if there are more execution groups, the task should stay in `RUNNING` state.

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskExecution.java:
```diff
@@ -209,6 +212,7 @@ public void testSimple(PipelineExecutionStrategy executionStrategy)
                     // assert that task result is produced
                     outputBufferConsumer.consume(300 + 200, ASSERT_WAIT_TIMEOUT);
                     outputBufferConsumer.assertBufferComplete(ASSERT_WAIT_TIMEOUT);
+                    assertEquals(taskStateMachine.getStateChange(RUNNING).get(10, SECONDS), FLUSHING);
```
shouldn't that be moved above `outputBufferConsumer.consume(300 + 200, ASSERT_WAIT_TIMEOUT);` to prevent race condition?

Does `outputBufferConsumer.assertBufferComplete(ASSERT_WAIT_TIMEOUT);` consume all output?

rohangarg replied:
Yes, `assertBufferComplete` makes sure that the buffer is complete. Added the assertions as discussed offline.

sopel39 commented on presto-tests/src/test/java/io/prestosql/execution/TestFlushingStageState.java:
```diff
@@ -0,0 +1,68 @@
+        waitForQueryState(queryRunner, queryId, RUNNING);
+
+        QueryInfo queryInfo = queryRunner.getCoordinator().getFullQueryInfo(queryId);
+        while (queryInfo.getOutputStage().get().getState() != FLUSHING) {
```
please use `assertEventually`

rohangarg replied:
fixed - also moved the creation of queryRunner to an unbounded `setup`. I think the test after setup should run within 30sec.

sopel39 commented on presto-main/src/test/java/io/prestosql/execution/TestSqlTaskManager.java:
```diff
@@ -110,21 +112,20 @@ public void testSimpleQuery()
                 results = sqlTaskManager.getTaskResults(taskId, OUT, results.getToken() + results.getSerializedPages().size(), DataSize.of(1, Unit.MEGABYTE)).get();
             }
-            assertEquals(results.isBufferComplete(), true);
+            assertTrue(results.isBufferComplete());
```
nit this should be separate commit

sopel39 commented:
merged, thanks!
